File: java/src/main/java/com/ellen/java/jmvp/base/BaseMvpFragment.java
package com.ellen.java.jmvp.base;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import javax.inject.Inject;

public abstract class BaseMvpFragment<P extends BasePresenter> extends Fragment {

    @Inject
    protected P mPresenter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        initMvp();

        //调用MVP的生命周期函数attach
        mPresenter.attachView();
        mPresenter.mView.attachPresenter(mPresenter);
        mPresenter.mModel.attachPresenter(mPresenter);

        mPresenter.attachLifeCycle(getLifecycle());
        return super.onCreateView(inflater, container, savedInstanceState);
    }

   protected abstract void initMvp();
}

File: app/src/main/java/com/ellen/androidarchitecture/kmvp/base/BaseMvpFragment.kt
package com.ellen.androidarchitecture.kmvp.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.ellen.androidarchitecture.kmvp.base.model.BaseModel
import javax.inject.Inject

abstract class BaseMvpFragment<P : BasePresenter<BaseModel,BaseView>> : Fragment() {
    @Inject
    protected lateinit var mPresenter:P

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        initMVP()

        //调用MVP的绑定attach周期函数
        mPresenter.attachedByView()
        mPresenter.mView.attachedByPresenter()
        mPresenter.mModel.attachedByPresenter()

        //让Presenter感知生命周期，以防止内存泄漏
        mPresenter.attachLifecycle(lifecycle)
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    abstract fun initMVP()
}